lib/srv/alpnproxy/kube.go

```diff
@@ -66,13 +98,94 @@ func (m *KubeMiddleware) OverwriteClientCerts(req *http.Request) ([]tls.Certific
 		return nil, trace.BadParameter("expect a TLS request")
 	}
 
+	m.certsMu.RLock()
 	cert, ok := m.certs[req.TLS.ServerName]
+	m.certsMu.RUnlock()
 	if !ok {
 		return nil, trace.NotFound("no client cert found for %v", req.TLS.ServerName)
 	}
+
+	newCert, err := m.ensureCertValidity(req.Context(), cert, req.TLS.ServerName)
```

Comment:
since the goroutine already updates the certificate in the map, we can skip it here and `cert.err:=m.ensureCert...`.

If the cert is valid, ensureCertValidity returns the cert itself